src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -341,13 +489,77 @@ internal bool StartBufferingAndTryToReadInStreamErrorPropertyValue(out ODataErro
             }
         }
 
+        /// <summary>
+        /// Asynchronously puts the reader into the state where it buffers read nodes.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        internal async Task StartBufferingAsync()
+        {
+            Debug.Assert(!this.isBuffering, "Buffering is already turned on. Must not call StartBuffering again.");
+            this.AssertAsynchronous();
+
+            if (this.bufferedNodesHead == null)
+            {
+                // capture the current state of the reader as the first item in the buffer (if there are none)
+                object value = await this.asyncInnerReader.GetValueAsync()
+                    .ConfigureAwait(false);
+                this.bufferedNodesHead = new BufferedNode(this.asyncInnerReader.NodeType, value);
+            }
+            else
+            {
+                this.removeOnNextRead = false;
+            }
+
+            Debug.Assert(this.bufferedNodesHead != null, "Expected at least the current node in the buffer when starting buffering.");
+
+            // Set the currentBufferedNode to the first node in the list; this means every time we start buffering we reset the
+            // position of the current buffered node since in general we don't know how far ahead we have read before and thus don't
+            // want to blindly continuing to read. The model is that with every call to StartBuffering you reset the position of the
+            // current node in the list and start reading through the buffer again.
+            if (this.currentBufferedNode == null)
+            {
+                this.currentBufferedNode = this.bufferedNodesHead;
+            }
+
+            this.isBuffering = true;
+        }
+
+        /// <summary>
+        /// A method to asynchronously detect whether the current property value represents an in-stream error.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains a tuple comprising of:
+        /// 1). A value of true if the current value is an in-stream error value; otherwise false.
+        /// 2). An <see cref="ODataError"/> instance that was read from the payload.</returns>
+        internal async Task<Tuple<bool, ODataError>> StartBufferingAndTryToReadInStreamErrorPropertyValueAsync()
+        {
+            this.AssertNotBuffering();
+            this.AssertAsynchronous();

```

Comment:
Why is this assertion here and not in other async methods? Is it necessary?

Reply:
I believe I added the assertion statement in other async methods as well but I'll verify